src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -1398,5 +1384,131 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+            foreach (ref readonly T item in items)
+            {
+                multiSet.Add(item);
+            }
+
+            return GetOrderedIndicesToRemoveFor(multiSet);
+        }
+
+        private ICollection<int>? GetOrderedIndicesToRemoveFor(MultiSet multiSet)
+        {
+            List<int>? indicesToRemove = null;
+            for (int i = 0; i < array!.Length; i++)
+            {
+                if (multiSet.TryRemove(array[i]))
+                {
+                    (indicesToRemove ??= new()).Add(i);
+                }
+            }
+
+            return indicesToRemove;
+        }
+
+        private SortedSet<int> FindOrderedIndicesToRemove(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
+        {
+            var indicesToRemove = new SortedSet<int>();
+            foreach (T item in items)
+            {
+                int index = -1;
+                do
+                {
+                    index = IndexOf(item, index + 1, equalityComparer);
+                } while (index >= 0 && !indicesToRemove.Add(index) && index < Length - 1);
+            }
+
+            return indicesToRemove;
+        }
+
+        private readonly struct MultiSet
+        {
+            private readonly Dictionary<NullableKeyWrapper, int> _dictionary;
+
+            public MultiSet(IEqualityComparer<T>? equalityComparer)
+            {
+                _dictionary = new Dictionary<NullableKeyWrapper, int>(new NullableKeyWrapperEqualityComparer(equalityComparer ?? EqualityComparer<T>.Default));
+            }
+
+            public void Add(T? item)
+            {
+#if NET6_0_OR_GREATER
+                ref int count = ref CollectionsMarshal.GetValueRefOrAddDefault(_dictionary, item, out _);
+                count++;
+#else
+            _dictionary[item] = _dictionary.TryGetValue(item, out int count) ? count + 1 : 1;
+#endif
+            }
+
+            public bool TryRemove(T? item)
+            {
+#if NET6_0_OR_GREATER
+                ref int count = ref CollectionsMarshal.GetValueRefOrNullRef(_dictionary, item);
+                if (Unsafe.IsNullRef(ref count) || count == 0)
+#else
+            if (!_dictionary.TryGetValue(item, out int count) || count == 0)
+#endif
+                {
+                    return false;
+                }
+
+#if NET6_0_OR_GREATER
+                count--;
+#else
+            _dictionary[item] = count - 1;
+#endif
+                return true;
+            }
+
+            private readonly struct NullableKeyWrapper
+            {
+                public readonly T? Key;
+
+                public static implicit operator NullableKeyWrapper(T? key)
+                {
+                    return new NullableKeyWrapper(key);
+                }
+
+                private NullableKeyWrapper(T? key)
+                {
+                    Key = key;
+                }
+            }
+
+            private class NullableKeyWrapperEqualityComparer : IEqualityComparer<NullableKeyWrapper>
```

Comment:
An opportunity for potential performance optimization is to do the following:
```C#
            private class NullableKeyWrapperEqualityComparer : IEqualityComparer<NullableKeyWrapper>
            {
                private readonly IEqualityComparer<T>? _keyComparer;

                public NullableKeyWrapperEqualityComparer(IEqualityComparer<T>? keyComparer)
                {
                    _keyComparer = keyComparer;
                }

                public int GetHashCode(NullableKeyWrapper key)
                {
                    return _keyComparer is null ? EqualityComparer<TKey>.Default.GetHashCode(key) : _keyComparer.GetHashCode(obj.Key);
                }

                public bool Equals(NullableKeyWrapper x, NullableKeyWrapper y)
                {
                    return  _keyComparer is null ? EqualityComparer<TKey>.Default.Equals(x.Key, y.Key) : _keyComparer.Equals(x.Key, y.Key);
                }
```

The JIT will optimize inlined calls to `EqualityComparer<TKey>.Default.Equals`/`GetHashCode` so this will save us one nested virtual call in the default case.

Reply:
Fixed.